Source/JavaScriptCore/jit/CCallHelpers.h

```diff
@@ -88,6 +89,50 @@ class CCallHelpers : public AssemblyHelpers {
         poke(GPRInfo::nonArgGPR0, POKE_ARGUMENT_OFFSET + argumentIndex - GPRInfo::numberOfArgumentRegisters);
     }
 
+    enum class ShuffleStatus : uint8_t {
+        ToMove,
+        BeingMoved,
+        Moved
+    };
+
+    template<typename RegType, size_t N, typename RegPair = std::pair<RegType, RegType>>
+    void emitShuffleMove(Vector<RegPair, N>& moves, Vector<ShuffleStatus, N>& status, unsigned index, RegType scratch)
+    {
+        status[index] = ShuffleStatus::BeingMoved;
+        for (unsigned i = 0; i < moves.size(); i ++) {
+            if (moves[i].first == moves[index].second) {
+                ASSERT(i != index);
```

Comment:
Is it ensured? I wonder if we should have early check like what we have in BBQJIT::emitShuffleMove.

Reply:
It should be ensured, in the existing implementation of `shuffleRegisters` we have:
```
for (unsigned i = 0; i < NumberOfRegisters; ++i) {
    if (sources[i] != destinations[i])
        pairs.append(std::make_pair(sources[i], destinations[i]));
}
```
so trivial moves are already checked for and shouldn't make it into the shuffle algorithm.